src/hotspot/share/compiler/tieredThresholdPolicy.cpp

```diff
@@ -507,16 +507,19 @@ nmethod* TieredThresholdPolicy::event(const methodHandle& method, const methodHa
     // method == inlinee if the event originated in the main method
     method_back_branch_event(method, inlinee, bci, comp_level, nm, THREAD);
     // Check if event led to a higher level OSR compilation
-    CompLevel expected_comp_level = comp_level;
+    CompLevel expected_comp_level = MIN2(CompLevel_full_optimization, static_cast<CompLevel>(comp_level + 1));
     if (!CompilationModeFlag::disable_intermediate() && inlinee->is_not_osr_compilable(expected_comp_level)) {
       // It's not possble to reach the expected level so fall back to simple.
       expected_comp_level = CompLevel_simple;
     }
-    nmethod* osr_nm = inlinee->lookup_osr_nmethod_for(bci, expected_comp_level, false);
-    assert(osr_nm == NULL || osr_nm->comp_level() >= expected_comp_level, "lookup_osr_nmethod_for is broken");
-    if (osr_nm != NULL) {
-      // Perform OSR with new nmethod
-      return osr_nm;
+    CompLevel max_osr_level = static_cast<CompLevel>(inlinee->highest_osr_comp_level());
+    if (max_osr_level >= expected_comp_level) { // fast check to avoid locking in a typical scenario
+      nmethod* osr_nm = inlinee->lookup_osr_nmethod_for(bci, expected_comp_level, false);
+      assert(osr_nm == NULL || osr_nm->comp_level() >= expected_comp_level, "lookup_osr_nmethod_for is broken");
+      if (osr_nm != NULL && osr_nm->comp_level() != comp_level) {
```

Comment:
Should you check  '> comp_level' here?  expected_comp_level could be CompLevel_simple and as result you can get our_nm with lower level than current comp_level.

Reply:
No we shouldn't. That's intentional. Sometimes we have to go to the lower level (3->1 or 2->1) when the method is not compilable at level 4.

Reply:
ok